Mathlib/Topology/Algebra/Group/Basic.lean

```diff
@@ -0,0 +1,2014 @@
+theorem TopologicalGroup.t2Space [T1Space G] : T2Space G := by
+  haveI := TopologicalGroup.t3Space G
+  infer_instance
+#align topological_group.t2_space TopologicalGroup.t2Space
+#align topological_add_group.t2_space TopologicalAddGroup.t2Space
+
+variable {G}
+variable (S : Subgroup G) [Subgroup.Normal S] [IsClosed (S : Set G)]
+
+@[to_additive]
+instance Subgroup.t3_quotient_of_isClosed (S : Subgroup G) [Subgroup.Normal S]
+    [IsClosed (S : Set G)] : T3Space (G ⧸ S) := by
+  suffices T1Space (G ⧸ S) by exact @TopologicalGroup.t3Space _ _ _ _ this
+  have hS : IsClosed (S : Set G) := inferInstance
+  rw [← QuotientGroup.ker_mk' S] at hS
+  exact TopologicalGroup.t1Space (G ⧸ S) (quotientMap_quotient_mk'.isClosed_preimage.mp hS)
+#align subgroup.t3_quotient_of_is_closed Subgroup.t3_quotient_of_isClosed
+#align add_subgroup.t3_quotient_of_is_closed AddSubgroup.t3_quotient_of_isClosed
+
+/-- A subgroup `S` of a topological group `G` acts on `G` properly discontinuously on the left, if
+it is discrete in the sense that `S ∩ K` is finite for all compact `K`. (See also
+`discrete_topology`.) -/
+@[to_additive
+  "A subgroup `S` of an additive topological group `G` acts on `G` properly
+  discontinuously on the left, if it is discrete in the sense that `S ∩ K` is finite for all compact
+  `K`. (See also `discrete_topology`."]
+theorem Subgroup.properlyDiscontinuousSMul_of_tendsto_cofinite (S : Subgroup G)
+    (hS : Tendsto S.subtype cofinite (cocompact G)) : ProperlyDiscontinuousSMul S G :=
+  { finite_disjoint_inter_image := by
+      intro K L hK hL
+      have H : Set.Finite _ := hS ((hL.prod hK).image continuous_div').compl_mem_cocompact
+      rw [preimage_compl, compl_compl] at H
+      convert H
+      ext x
+      simp only [image_smul, mem_setOf_eq, coeSubtype, mem_preimage, mem_image, Prod.exists]
+      exact Set.smul_inter_ne_empty_iff' }
+#align subgroup.properly_discontinuous_smul_of_tendsto_cofinite Subgroup.properlyDiscontinuousSMul_of_tendsto_cofinite
+#align add_subgroup.properly_discontinuous_vadd_of_tendsto_cofinite AddSubgroup.properlyDiscontinuousVAdd_of_tendsto_cofinite
+
+-- attribute [local semireducible] MulOpposite -- Porting note: doesn't work in Lean 4
+
+/-- A subgroup `S` of a topological group `G` acts on `G` properly discontinuously on the right, if
+it is discrete in the sense that `S ∩ K` is finite for all compact `K`. (See also
+`discrete_topology`.)
+
+If `G` is Hausdorff, this can be combined with `t2_space_of_properly_discontinuous_smul_of_t2_space`
+to show that the quotient group `G ⧸ S` is Hausdorff. -/
+@[to_additive
+  "A subgroup `S` of an additive topological group `G`
+  acts on `G` properly discontinuously on the right,
+  if it is discrete in the sense that `S ∩ K` is finite for all compact `K`.
+  (See also `discrete_topology`.)
+
+  If `G` is Hausdorff, this can be combined with
+  `t2_space_of_properly_discontinuous_vadd_of_t2_space`
+  to show that the quotient group `G ⧸ S` is Hausdorff."]
```

Comment:
```suggestion
If `G` is Hausdorff, this can be combined with `t2Space_of_properlyDiscontinuousSMul_of_t2Space`
to show that the quotient group `G ⧸ S` is Hausdorff. -/
@[to_additive
  "A subgroup `S` of an additive topological group `G`
  acts on `G` properly discontinuously on the right,
  if it is discrete in the sense that `S ∩ K` is finite for all compact `K`.
  (See also `DiscreteTopology`.)

  If `G` is Hausdorff, this can be combined with
  `t2Space_of_properlyDiscontinuousVAdd_of_t2Space`
  to show that the quotient group `G ⧸ S` is Hausdorff."]
```